### PR TITLE
PR 1 of 3: focal CN file preparation

### PR DESCRIPTION
**Pull request by cbethell**
### Purpose/implementation

This is the first of tentatively two PRs written to take a seg file with copy number information and prepare a focal copy number file for consumption in the oncoprint (#6) and co-occurrence/mutual exclusivity modules (#13).

### Issue

This PR addresses issue #186.

### Directions for reviewers

1. In the `label` column of the output file, I included the values: `Amplification`, `Hem_Deletion`, and `Hom_Deletion`. Would it be more useful to specify Diploid vs Gain here? 
2. Does this analysis seem to adequately produce a focal CN file that can be used in downstream analyses? In other words, are there any necessary steps that may be missing from this workflow? 

### Results

The `01-prepare-cn-file.R` script outputs a tsv file here:
- `analyses/focal-cn-file-preparation/results/annotated_cn.tsv`

It is the following tabular format:

| gene_symbol | biospecimen_id | label |  copy_number |
| ------------- | --------------- | ---- | ------------ |
| YFG1 | BS_XXXXXXXX |Hem_Deletion | 1 |
| ... | ... | ... | ... |
| ... | ... | ... | ... |
| ... | ... | ... | ... |

#### Docker and continuous integration

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile. 
- [x] This analysis has been added to continuous integration.

#### PR Checklist
* [x] Run a linter
* [x] Set the seed- Not needed.
* [x] Comments and/or documentation up to date 
* [x] Double check your paths
* [x] Spell check any Rmd file or md file
* [x] Restart R and run all notebooks fresh and save

### CODE REVIEW

File: analyses/focal-cn-file-preparation/01-prepare-cn-file.R [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+# names(seg_no_xy) <- c("chr", "start", "end", "markers", "CN", "Model")
+
+# Make seg data.frame a GRanges object 
+seg_gr <- seg_no_xy %>%
+  GenomicRanges::makeGRangesFromDataFrame(keep.extra.columns = TRUE,
+                                          starts.in.df.are.0based = FALSE)
+
+# Define the annotations for the hg38 genome
+annotations <- annotatr::build_annotations(genome = "hg38",
+                                           annotations = "hg38_basicgenes")
```
Comment (jashapiro) [inline analyses/focal-cn-file-preparation/01-prepare-cn-file.R]:
```suggestion
                                           annotations = "hg38_genes_exons")
```

This should speed up this and the next step

File: analyses/focal-cn-file-preparation/01-prepare-cn-file.R [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+opt <- optparse::parse_args(opt_parser)
+
+#### Directories and Files -----------------------------------------------------
+
+results_dir <- "results"
+if (!dir.exists(results_dir)) {
+  dir.create(results_dir)
+}
+
+seg_file <-
```
[the remark this answers is not in the document]
Comment (cbethell) [reply analyses/focal-cn-file-preparation/01-prepare-cn-file.R]: Good point.

**Comment by cbethell** [comment]
> Looks good, with the caveat that I wasn't actually able to run it locally due to memory issues (I assume).

@jashapiro Yes, I used the subset file to run it locally and produce the current tsv file.

**Comment by jashapiro** [comment]
This looks good, but doing some spot checking, I found a bit of strangeness that might bear investigating/discussion. Specifically, the following two lines: 
```
LTBP1	BS_XEVMEYFS	Gain	3
LTBP1	BS_XEVMEYFS	Hom_Deletion	0
```
The same gene, with both deletion and gain. in the same sample.
I assume this means that part of the gene is deleted and part is amplified, but I am worried.

A quick check through the file reveals this happens about 30,000 times, which isn't insignificant. 
```
multi_hits <- copy_number %>% 
  dplyr::group_by(biospecimen_id, gene_symbol) %>% 
  tally() %>% 
  filter(n > 1) 
nrow(multi_hits)
## [1] 29814

summary(as.factor(multis$n))
##     2     3     4     5     6     7 
## 27422  2116   220    39    14     3 

```

I don't think this is a problem with the scripts as they stand, but should be noted.

**Comment by jaclyn-taroni** [comment]
Something to keep an eye on for sure!

**Pull request merged**